File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14638/[C#][C++]-Story Creator Sample Application for Windows Phone 7 (CSWP7AzureVideoStory)/C# and C++/C#,C++/CSWP7AzureVideoStory/VideoStoryCreator/Transitions/TransitionFactory.cs
/****************************** Module Header ******************************\
* Module Name:	TransitionFactory.cs
* Project:		VideoStoryCreator
* Copyright (c) Microsoft Corporation.
* 
* A factory class used to create transitions and their designers.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Collections.Generic;
using System.Windows.Controls;

namespace VideoStoryCreator.Transitions
{
    public class TransitionFactory
    {
        private static Dictionary<string, Type> _transitionNameTypes = new Dictionary<string, Type>();
        private static Dictionary<string, Type> _transitionNameDesigners = new Dictionary<string, Type>();
        private static List<string> _transitionNames = new List<string>();

        static TransitionFactory()
        {
            _transitionNameTypes.Add("Fade Transition", typeof(FadeTransition));
            _transitionNames.Add("Fade Transition");

            _transitionNameTypes.Add("Fly In Transition", typeof(FlyInTransition));
            _transitionNameDesigners.Add("Fly In Transition", typeof(FlyInTransition_Design));
            _transitionNames.Add("Fly In Transition");

            // Register additional transitions here...
        }

        private TransitionFactory()
        {
        }

        /// <summary>
        /// Create a transition based on its name.
        /// </summary>
        /// <param name="name">The transition's name.</param>
        /// <returns>An ITransition object.</returns>
        public static ITransition CreateTransition(string name)
        {
            if (_transitionNameTypes.ContainsKey(name))
            {
                Type transitionType = _transitionNameTypes[name];
                try
                {
                    return Activator.CreateInstance(transitionType) as ITransition;
                }
                catch
                {
                    // TODO: Should we throw the exception or simply return null?
                }
            }
            return null;
        }

        /// <summary>
        /// For now, by default we create a fade transition.
        /// </summary>
        public static ITransition CreateDefaultTransition()
        {
            return new FadeTransition();
        }

        /// <summary>
        /// Create a transition designer based on its name.
        /// </summary>
        /// <param name="name">The transition's name.</param>
        /// <returns>A UserContro. All designers must inherite from UserControl.</returns>
        public static UserControl CreateTransitionDesigner(string name)
        {
            if (_transitionNameDesigners.ContainsKey(name) && _transitionNameDesigners[name] != null)
            {
                Type transitionDeginerType = _transitionNameDesigners[name];
                try
                {
                    return Activator.CreateInstance(transitionDeginerType) as UserControl;
                }
                catch
                {
                    // Todo: Should we throw the exception or simply return null?
                }
            }
            return null;
        }

        /// <summary>
        /// Return a list of transition names.
        /// </summary>
        public static List<string> AvailableTransitions
        {
            get { return _transitionNames; }
        }
    }
}

File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14992/[VB][C++]-Story Creator Sample Application for Windows Phone 7 (VBWPAzureVideoStory)/C++ and VB/C++,VB/VBWPAzureVideoStory/VideoStoryCreator/Transitions/TransitionFactory.vb
'********************************* Module Header *********************************\
' Module Name: TransitionFactory.vb
' Project: VideoStoryCreator
' Copyright (c) Microsoft Corporation.
' 
' A factory class used to create transitions and their designers.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************/



Public Class TransitionFactory
    Private Shared _transitionNameTypes As New Dictionary(Of String, Type)()
    Private Shared _transitionNameDesigners As New Dictionary(Of String, Type)()
    Private Shared _transitionNames As New List(Of String)()

    Shared Sub New()
        _transitionNameTypes.Add("Fade Transition", GetType(FadeTransition))
        _transitionNames.Add("Fade Transition")

        _transitionNameTypes.Add("Fly In Transition", GetType(FlyInTransition))
        _transitionNameDesigners.Add("Fly In Transition", GetType(FlyInTransition_Design))

        ' Register additional transitions here...
        _transitionNames.Add("Fly In Transition")
    End Sub

    Private Sub New()
    End Sub

    ''' <summary>
    ''' Create a transition based on its name.
    ''' </summary>
    ''' <param name="name">The transition's name.</param>
    ''' <returns>An ITransition object.</returns>
    Public Shared Function CreateTransition(name As String) As ITransition
        If _transitionNameTypes.ContainsKey(name) Then
            Dim transitionType As Type = _transitionNameTypes(name)
            Try
                Return TryCast(Activator.CreateInstance(transitionType), ITransition)
                ' TODO: Should we throw the exception or simply return null?
            Catch
            End Try
        End If
        Return Nothing
    End Function

    ''' <summary>
    ''' For now, by default we create a fade transition.
    ''' </summary>
    Public Shared Function CreateDefaultTransition() As ITransition
        Return New FadeTransition()
    End Function

    ''' <summary>
    ''' Create a transition designer based on its name.
    ''' </summary>
    ''' <param name="name">The transition's name.</param>
    ''' <returns>UserControl. All designers must inherite from UserControl.</returns>
    Public Shared Function CreateTransitionDesigner(name As String) As UserControl
        If _transitionNameDesigners.ContainsKey(name) AndAlso _transitionNameDesigners(name) IsNot Nothing Then
            Dim transitionDeginerType As Type = _transitionNameDesigners(name)
            Try
                Return TryCast(Activator.CreateInstance(transitionDeginerType), UserControl)
                ' Todo: Should we throw the exception or simply return null?
            Catch
            End Try
        End If
        Return Nothing
    End Function

    ''' <summary>
    ''' Return a list of transition names.
    ''' </summary>
    Public Shared ReadOnly Property AvailableTransitions() As List(Of String)
        Get
            Return _transitionNames
        End Get
    End Property
End Class
